[quant][graphmode][fx][fix] Fix observer insert logic for ops that inherits quantization parameters for input

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45539 [quant][graphmode][fx][eagermode] Support sigmoid/hardsigmoid/tanh in eager and fx graph mode
* #45538 [quant] Add FixedQParamsFakeQuantize module
* **#45473 [quant][graphmode][fx][fix] Fix observer insert logic for ops that inherits quantization parameters for input**
* #45292 [quant][graphmode][fx] Merge all quantization mode
* #45343 [quant] Use PlaceholderObserver as default dynamic quant observer

Summary:
Currently we don't insert observer for the output, but we actually need to
use input observer for the output to be consistent with the transformation
we are doing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23981312](https://our.internmc.facebook.com/intern/diff/D23981312)

jerryzh168 committed:
[quant][graphmode][fx][fix] Fix observer insert logic for ops that inherits quantization parameters for input

Summary:
Currently we don't insert observer for the output, but we actually need to
use input observer for the output to be consistent with the transformation
we are doing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fix] Fix observer insert logic for ops that inherits quantization parameters for input"


Summary:
Currently we don't insert observer for the output, but we actually need to
use input observer for the output to be consistent with the transformation
we are doing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23981312](https://our.internmc.facebook.com/intern/diff/D23981312)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fix] Fix observer insert logic for ops that inherits quantization parameters for input"


Summary:
Currently we don't insert observer for the output, but we actually need to
use input observer for the output to be consistent with the transformation
we are doing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23981312](https://our.internmc.facebook.com/intern/diff/D23981312)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fix] Fix observer insert logic for ops that inherits quantization parameters for input"


Summary:
Currently we don't insert observer for the output, but we actually need to
use input observer for the output to be consistent with the transformation
we are doing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23981312](https://our.internmc.facebook.com/intern/diff/D23981312)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fix] Fix observer insert logic for ops that inherits quantization parameters for input"


Summary:
Currently we don't insert observer for the output, but we actually need to
use input observer for the output to be consistent with the transformation
we are doing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23981312](https://our.internmc.facebook.com/intern/diff/D23981312)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fix] Fix observer insert logic for ops that inherits quantization parameters for input"


Summary:
Currently we don't insert observer for the output, but we actually need to
use input observer for the output to be consistent with the transformation
we are doing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23981312](https://our.internmc.facebook.com/intern/diff/D23981312)

[ghstack-poisoned]

supriyar commented on torch/quantization/fx/quantize.py:
```diff
@@ -374,15 +376,24 @@ def load_arg(a):
+                # insert a call to activation_post_process module
+                def insert_observer_call(node, observer_name):
                     env[node.name] = observed_graph.create_node('call_module', observer_name, (load_arg(node),), {})
+                    activation_post_process_name_map[node.name] = observer_name
```
Shouldn't this be `self.activation_post_process`?

jerryzh168 replied:
oh this is storing the names of activation_post_process modules, since we need this information when we insert observer calls, and it's not stored anywhere else.

jerryzh168 replied:
this can be a local variable since it's just needed when we process the graph

supriyar commented on torch/quantization/fx/quantize.py:
```diff
@@ -417,7 +428,18 @@ def insert_observer(node, observer, device):
+                    # use the same observer as input
+                    observed_input = node.args[0]
+                    if observed_input.name in activation_post_process_name_map:
+                        print('inserting observer of input', observed_input.target)
```
Extra print statement?

supriyar reviewed (approved):
Left couple comments that need to be addressed.

jerryzh168 committed:
Update on "[quant][graphmode][fx][fix] Fix observer insert logic for ops that inherits quantization parameters for input"


Summary:
Currently we don't insert observer for the output, but we actually need to
use input observer for the output to be consistent with the transformation
we are doing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23981312](https://our.internmc.facebook.com/intern/diff/D23981312)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fix] Fix observer insert logic for ops that inherits quantization parameters for input"


Summary:
Currently we don't insert observer for the output, but we actually need to
use input observer for the output to be consistent with the transformation
we are doing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23981312](https://our.internmc.facebook.com/intern/diff/D23981312)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fix] Fix observer insert logic for ops that inherits quantization parameters for input"


Summary:
Currently we don't insert observer for the output, but we actually need to
use input observer for the output to be consistent with the transformation
we are doing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23981312](https://our.internmc.facebook.com/intern/diff/D23981312)

[ghstack-poisoned]

jerryzh168 commented:
this is not needed anymore, we plan to observe the output for all the ops that inherits quantization parameters (e.g. average pool) and ops that currently have fixed quantization parameters (e.g. sigmoid).
